contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,104 @@
+
+In the code snippet below, `console.log(price)` would return undefined as the asynchronous `retrievePrice()` function has not finished executing.
+
+```
+var price;
+price = retrievePrice("https://www.amazon.com/gp/product/B009S331VU/");
+console.log(price);
+
+```
+
+As such, it can be difficult to predict how your Javascript code executes as asynchronous methods such as the above would not necessarily execute in the sequence that is desired. 
+
+This is where Promises would come in handy.
+
+
+## Introduction
+Promises in Javscript behave the same way as Promises do in real life. It is a proxy for a value that is not known at the time of Promise creation. It has 3 states: pending, fulfilled and rejected. Promises provide the power to control the execution flow of your program by allowing the user to wait for a promise to be resolved before carrying on with the rest of the program execution.
+
+## Benefits
+
+- Avoid callback hell
+
+![Callback Hell](http://blog.mclain.ca/assets/images/callbackhell.png)
+
+In Javascript, callback functions are usually used to perform some action on a value returned from another function. However, this may result in deeply nested code as shown above, especially if there are many asynchronous method calls. When using promises, the nesting is kept at a minimum and results in cleaner and readable code as shown below. 
+
+- Chain multiple asynchronous methods
+
+Calling asynchronous code can be tedious, especially if a later function relies on data that is produced by an earlier function. By using Promises, these asynchronous methods can be chained and variables will be correctly set instead of being undefined when the variable needs to be used.
+
+- Control execution flow
+
+As mentioned earlier, calling asynchronous methods in Javascript may not necessarily result in a sequential execution flow. However, it is possible to ensure that methods get called in the order that they are written in, as long as they are wrapped in a Promise.
+
+
+## How
+
+Promises are created in Javascript by using the `new` constructor as follows:
+
+```
+
+var myFirstPromise = new Promise(function(resolve, reject) {
+    // insert stuff to do here
+    // The promise is currently 'pending'
+})
+
+```
+
+The Promise constructor takes function that has two parameters: resolve and reject. These two parameters are functions which will change the state of the promise. They can be called with or without a parameter as follows:
+
+```
+
+resolve(someValue);
+reject(someValue);
+resolve();
+reject();
+
+```
+
+## Handling
+
+As promises have 2 states, there are 2 functions available to deal with these 2 scenarios. 
```

Comment:
Earlier in your text, you said that Promises have 3 states. Do be careful not to confuse readers. Perhaps you were referring to "2 completed states" instead.

Reply:
Good catch.